source/extensions/filters/http/compressor/compressor_filter.cc

```diff
@@ -0,0 +1,24 @@
+#include "extensions/filters/http/compressor/compressor_filter.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Compressor {
+
+CompressorFilterConfig::CompressorFilterConfig(
+    const envoy::extensions::filters::http::compressor::v3::Compressor& generic_compressor,
+    const std::string& stats_prefix, Stats::Scope& scope, Runtime::Loader& runtime,
+    CompressorFactoryPtr compressor_factory)
+    : Common::Compressors::CompressorFilterConfig(
+          generic_compressor, stats_prefix + compressor_factory->statsPrefix(), scope, runtime,
```

Comment:
I think it would be valuable to insert "compressor" here e.g., `stats_prefix + "compressor." + compressor_factory->statsPrefix()`.

Reply:
Actually in thinking more about it (in conjunction with my suggestion that compressors become an extension point) I think we should have three levels of namespaces:

1. The compressor filter stats
2. The general compressor stats
3. The specific compressor stats

For example, we might have something like:
```
http_filter.compressor_filter.response_compressed.counter
http_filter.compressor_filter.compressor.buffer_compressed_failure.counter
http_filter.compressor_filter.compressor.gzip.window_bits_exceeded.counter
```

Reply:
What if there are two compressor filters configured to use different encodings for different content types? I guess `http_filter.compressor_filter.compressor.buffer_compressed_failure.counter` will become indistinguishable between the two. And `.compressor_filter.compressor.` sounds like a tautology.

Reply:
Good point on the former, although that problem still holds with the current scheme. Perhaps we need an identifier per filter similar to listeners, there might already be something in place, let me take a look.

For the latter, I agree it is a tautology, but I still think there is value in the separate namespaces, one if for stats pertinent to the compressor filter and the other is for stats pertinent to the compressor the filter is using. Any suggestions on better naming? Or do you think it is overkill to give the compressor itself stats, and all the stats should live on the filter?

Reply:
I'm trying to put on a user's hat. As a user I'd probably cared about a filter's stats only. If there're stats specific to a compressor they would live in `compression.compressor.*` and `compression.decompressor.*` namespaces.

Reply:
Right, they would be under that, but because we are passing the stats_prefix they would be prefixed under the http filter. So we would know that those stats refer to the compressor used by the compressor http filter. So for instance:

`http_filter.compressor_filter.compression.gzip.compressor.window_bits_exceeded` would tell us about the times window_bit_exceeded happened as used by the http compressor filter. However, you are right that if two filters used the gzip compressor we could not discern between the two

Reply:
Oops, missed.

Ok, I've moved compressor's stats to its own namespace. Now they look like
```
http.ingress_http.compressor.gzip.compressor.compressed: 3
http.ingress_http.compressor.gzip.compressor.content_length_too_small: 0
http.ingress_http.compressor.gzip.compressor.header_compressor_overshadowed: 0
http.ingress_http.compressor.gzip.compressor.header_compressor_used: 3
http.ingress_http.compressor.gzip.compressor.header_gzip: 3
http.ingress_http.compressor.gzip.compressor.header_identity: 0
http.ingress_http.compressor.gzip.compressor.header_not_valid: 0
http.ingress_http.compressor.gzip.compressor.header_wildcard: 0
http.ingress_http.compressor.gzip.compressor.no_accept_header: 0
http.ingress_http.compressor.gzip.compressor.not_compressed: 0
http.ingress_http.compressor.gzip.compressor.not_compressed_etag: 0
http.ingress_http.compressor.gzip.compressor.total_compressed_bytes: 624
http.ingress_http.compressor.gzip.compressor.total_uncompressed_bytes: 768

```

The first `compressor` relates a stat to the compressor filter and the second indicates compression direction: inflating or deflating.

Reply:
Nice, so now a user knows that those are the compressor filter stats that is using a gzip compressor. My only remaining doubt is that if we put everything under that tree:
1. Where would stats owned by the compressor class go? If they go under the same tree then it would be hard to know which stats are coming from the filter and which are coming from the compressor itself at first glance. We could have the filter stats not have the second compressor prefix, and have stats exposed by the compressor itself be under the second compressor prefix:
```
http.ingress_http.compressor.gzip.compressed: 3
# I know this stat doesn't exist, nor any stats "owned" by the compressor itself, but what if there were?
http.ingress_http.compressor.gzip.compressor.window_bits_exceeded: 10
```
2. Do you think people might configure compressor filters in the same chain with the same compressor library but different parameters?

Reply:
1. Right. We could remove the second `compressor` from `GzipCompressorFactory::statsPrefix()`'s return value in this PR. Then if compressors ever provide their own stats like `window_bits_exceeded` they'll need to add their own prefix (e.g. `compressor.`) to whatever prefix is given by the context (e.g. `http.ingress_http.compressor.gzip.`). WDYT?
2. Yes, this is natural that different Content-Types may require different compression strategies. After all the strategies themselves were introduced to improve compression for certain content types, i.e. RLE works better for PNG images.

Reply:
1. sgtm!
2. @rojkov I mostly meant installing two compressor filters, both with the gzip compressor, but with different parameters. Because in that case the stats for both installations would still get mixed.

Reply:
1. done :)
2. So did I. Consider this config
```
          http_filters:
          - typed_config:
              "@type": type.googleapis.com/envoy.extensions.filters.http.compressor.v3.Compressor
              content_type:
              - image/png
              compressor_library:
                typed_config:
                  "@type": type.googleapis.com/envoy.extensions.compression.gzip.compressor.v3.Gzip
                  compression_strategy: RLE
          - typed_config:
              "@type": type.googleapis.com/envoy.extensions.filters.http.compressor.v3.Compressor
              content_type:
              - text/html
              - application/javascript
              compressor_library:
                typed_config:
                  "@type": type.googleapis.com/envoy.extensions.compression.gzip.compressor.v3.Gzip
                  compression_strategy: DEFAULT_STRATEGY
```
Even though the same gzip compressor is used for both PNG and text contents it's configured to use different strategies for them as PNG benefits from RLE while HTML compression may underperform with it.

Perhaps we could add an optional `ID` field to compressor config to be added to the stats prefix just to distinguish different filters in stats. But so far users haven't requested it for the existing `gzip` filter.

Reply:
Gotcha! Sorry I misunderstood strategy for using a different compression library altogether. Yes, that is the exact scenario I was thinking of.

We could use the filter's name field to add to the stat's tree to differentiate them. After https://github.com/envoyproxy/envoy/issues/9466 was closed the name can be an arbitrary string. So we could have:
```
http.ingress_http.compressor.png.gzip.
http.ingress_http.compressor.web.gzip.
```

wdyt @rojkov?

Reply:
Yes, that sounds good. In this case suffixing with `gzip.` may be redundant even, unless no name is configured.
I presume this will require updating `createFilterFactoryFromProto()`'s signature to accommodate filter names. Right?

Reply:
Hmm interesting point about using only the name if present, and the generic compression library's prefix otherwise. I still like the compression library name there, specially if we eventually have library-specific stats.

> I presume this will require updating createFilterFactoryFromProto()'s signature to accommodate filter names. Right?

Dang, for some reason I thought that we had access to the filter name in the factory function after #9587. Yep, that would be the way to do it. Originally I thought about only changing `createFilterFactoryFromProto` and storing the name, but the factory itself is static. I wanted to prevent changing the `createFilterFactoryFromProtoTyped()` signature to avoid churn in all filters, but I don't see a way around it... Unless we add a "name" field to the compressor filter config.

@htuch and @kyessenov what do you think? Now that the filter name can be arbitrary, we would like access to it in the compressor filter in order to distinguish stats trees between different installations of the filter with the same compression libraries.